### ISSUE TITLE
Race condition with snapshot listener started before batched write finished in backend

### ISSUE THREAD

**Issue by SamyPesse** [firebase/firebase-js-sdk#5768]
### Describe your environment

  * Operating System version: Mac OS X
  * Browser version: Chrome Latest
  * Firebase SDK version: v9
  * Firebase Product: firestore

### Describe the problem

We are currently observing a "race-condition" issue:

Imagine a docA that is referencing a docB. And an application that needs to listen to the ref of docA to locate docB and then listen to docB.

If a batch commit updates both docA and docB, we can end-up with the following situation:

1. batch commit starts
2. docA is updated in commit
3. docB is updated in commit
4. batch commit is applied locally
5. Listener of docA is called with the new document, `snapshot.metadata` is `{ hasPendingWrites: true, fromCache: true }`
6. We start a `onSnapshot` listener for docB using the reference from the docA data from the step 5.
5. Listener of docB is called with the new document, `snapshot.metadata` is `{ hasPendingWrites: true, fromCache: true }`
8. batch commit is applied in the backend
9. :boom: Listener of docB is called with a non existing document and `{ hasPendingWrites: false, fromCache: false }`

Basically it looks like when we've started listening to the second document (`docB`) before the commit was finished in the backend, it is being triggered with a non existing document.
I can imagine it's because the backend is sending the data (in that case "document not found") before it got the data from the commit.

But shouldn't it be handled by the JS sdk since the sdk knows that the document has pending writes ?

#### Steps to reproduce:

I'm looking at creating a codesandbox to reproduce this.

**Comment by thebrianchen** [comment]
@SamyPesse Thanks for filing a detailed issue. Just to double check my understanding and assumptions: 
- The two document listeners are added before the batch commit is made.
- The update to docA changes the docB reference.
- docB's values are updated but not deleted.

> Basically it looks like when we've started listening to the second document (docB) before the commit was finished in the backend, it is being triggered with a non existing document.

Just to double check here, are you setting a new listener inside the write batch code? 

I think this might be related to an [issue](https://github.com/firebase/firebase-ios-sdk/issues/8944#issuecomment-967286847) on the iOS SDK. If you have a repro, that would be immensely helpful in diagnosing the issue. Thanks!
